src/storage/src/hummock/time_travel_version_cache.rs

```diff
@@ -62,11 +40,25 @@ impl SimpleTimeTravelVersionCacheInner {
         Self { cache }
     }
 
-    fn get(&self, epoch: &HummockEpoch) -> Option<PinnedVersion> {
-        self.cache.get(epoch)
-    }
-
-    fn add(&mut self, epoch: HummockEpoch, version: PinnedVersion) {
-        self.cache.insert(epoch, version);
+    pub async fn get_or_insert(
+        &self,
+        table_id: u32,
+        epoch: HummockEpoch,
+        fetch: impl Future<Output = HummockResult<PinnedVersion>> + Send + 'static,
+    ) -> HummockResult<PinnedVersion> {
+        let result = self
+            .cache
+            .entry((table_id, epoch))
+            .or_insert_with(|| fetch.boxed().shared())
+            .value()
+            .clone()
+            .await;
+        if result.is_err() {
```

Comment:
In terms of error handling, we can handle the the case of error in an atomic way with [or_insert_with_if](https://docs.rs/moka/latest/moka/sync/struct.OwnedKeyEntrySelector.html#method.or_insert_with_if) of the entry. 

Basically we want to reset the cache entry with a new future when we know that, first, the original future is ready, and second, the output of the original future is error. We can use `if let Some(result) = future.peek() && result.is_err()` to detect the case.

Reply:
Thanks